File: src/main/java/baekjoon/dp/카드_구매하기.java
package baekjoon.dp;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

/**
 * https://www.acmicpc.net/problem/11052
 *
 * - fn(n) = n 개 카드를 구매할 때 드는 비용의 최대값
 *
 * - 예를 들어 카드를 4개 구매해야 한다고 하면
 *   - 4개짜리 카드덱을 구매했을 때의 가격
 *   - 3개짜리 카드덱을 구매했을 때 가격 + 카드 1개를 구하기 위해 구매한 카드덱 가격의 최대 값
 *   - 2개짜리 카드덱을 구매했을 때 가격 + 카드 2개를 구하기 위해 구매한 카드덱 가격의 최대 값
 *   - 1개짜리 카드덱을 구매했을 때 가격 + 카드 3개를 구하기 위해 구매한 카드덱 가격의 최대 값
 * - 즉, 위의 결과에서 최대값을 구하면 된다
 *   - max(fn(n-i) + pricesZ[i]))
 *
 * - 시간 복잡도는 O(n^2) 가 된다
 *   - n을 받으면 1~n 까지 반복(O(n))을 돌고, 이것을 n 이 0이 될때까지(O(n)) 하니 O(n^2) 이다
 */
public class 카드_구매하기 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        카드_구매하기 fn = new 카드_구매하기(Arrays.stream(br.readLine().split(" ", n))
                .mapToInt(Integer::parseInt)
                .toArray());
        System.out.println(fn.bottomUp(n));
    }

    int[] prices;
    public 카드_구매하기(int[] args) {
        this.prices = new int[args.length + 1];
        System.arraycopy(args, 0, this.prices, 1, args.length);
    }

    /*
    array[1] = 카드 1개 구매할 때 최대 비용
    array[2] = 카드 2개 구매할 때 최대 비용
    array[3] = 카드 3개 구매할 때 최대 비용
    ...
     */
    public int bottomUp(int n) {
        int[] dp = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            dp[i] = prices[i];

            for (int j = 1; j < i; j++) {
                dp[i] = Math.max(dp[i], prices[j] + dp[i - j]);
            }
        }

        return dp[n];
    }


    // 매 함수마다 배열이 따로 생성되고
    public int topDown(int n) {
        if(n == 0) {
            return 0;
        }

        int[] array= new int[n + 1];

        int max = 0;
        for (int i = 1; i <= n; i++) {
            array[i] = prices[i] + topDown(n - i);

            if(array[i] > max) {
                max = array[i];
            }
        }

        return max;
    }
}

File: src/main/java/baekjoon/dp/basic/카드 구매하기.kt
package baekjoon.dp.basic

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

/**
 * https://www.acmicpc.net/problem/11052
 *
 * f(n) = 카드 n개를 구매하기 위해 드는 최대값
 * p(n) = n개 카드가 들어있는 덱의 가격
 *
 * 4개의 카드를 구매해야 한다고 했을 떄,
 * 1개 덱 카드를 선택했을 경우 : p(1) + f(n-1)
 * 2개 덱 카드를 선택했을 경우 : p(2) + f(n-2)
 * 3개 덱 카드를 선택했을 경우 : p(3) + f(n-3)
 * 4개 덱 카드를 선택했을 경우 : p(4) + f(n-4)
 *
 * 이 값들의 max를 구하면 된다 : max(prices(i) + fn(n-i))
 */
fun main(args: Array<String>) {
    println(`카드 구매하기`().solution(4, intArrayOf(1,5,6,7)) == 10)
    println(`카드 구매하기`().solution(5, intArrayOf(10,9,8,7,6)) == 50)
    println(`카드 구매하기`().solution(10, intArrayOf(1,1,2,3,5,8,13,21,34,55)) == 55)

    println(`카드 구매하기 topDown`(4).solution(4, intArrayOf(1,5,6,7)) == 10)
    println(`카드 구매하기 topDown`(5).solution(5, intArrayOf(10,9,8,7,6)) == 50)
    println(`카드 구매하기 topDown`(10).solution(10, intArrayOf(1,1,2,3,5,8,13,21,34,55)) == 55)

    try {
        val fn = `카드 구매하기`()
        val br = BufferedReader(InputStreamReader(System.`in`))
        val count = br.readLine().toInt()
        val prices = br.readLine().split(" ").stream()
            .mapToInt { it.toInt() }
            .toArray()
        println(fn.solution(count, prices))
    } catch (e: IOException) {/* */}
}

/**
 * 1개일 경우 최대 금액(prices[1]) 부터 n 개일 경우 최대 금액까지 쌓아 올리기
 */
class `카드 구매하기` {
    // bottomUp
    fun solution(n: Int, prices: IntArray): Int {
        val dp = IntArray(n + 1)

        for (i in 1..n) {
            for (j in 1..i) {
                dp[i] = dp[i].coerceAtLeast(prices[j - 1] + dp[i - j])
            }
        }

        return dp[n]
    }
}

class `카드 구매하기 topDown`(n: Int) {
    // topDown
    private val dp = IntArray(n + 1)
    fun solution(n: Int, prices: IntArray): Int {
        if (dp[n] > 0) {
            return dp[n]
        }

        for (i in 1..n) {
            dp[n] = dp[n].coerceAtLeast(prices[i - 1] + solution(n - i, prices))
        }

        return dp[n]
    }
}

